libqtile/widget/bluetooth.py

```diff
@@ -0,0 +1,114 @@
+BLUEZ_DEVICE = 'org.bluez.Device1'
+BLUEZ_PROPERTIES = 'org.freedesktop.DBus.Properties'
+
+class Bluetooth(base._TextBox):
+    """
+    Displays bluetooth status or connected device.
+
+    Uses dbus to communicate with the system bus.
+    """
+
+    orientations = base.ORIENTATION_HORIZONTAL
+    defaults = [
+        ('update_interval', 1, 'The update interval.'),
+        ('format', '{status}', 'Display format'),
+        ('hci', '/dev_XX_XX_XX_XX_XX_XX', 'hci0 device path, can be found with d-feet or similar dbus explorer.')
+    ]
+
+    def __init__(self, **config):
+        base._TextBox.__init__(self, '', **config)
+        self.add_defaults(Bluetooth.defaults)
+
+        self.hci = config['hci']
+
+    async def _config_async(self):
+        # set initial values
+        self.powered = await self._init_adapter()
+        self.connected, self.device = await self._init_device()
+
+        self.update_text()
+
+    async def _init_adapter(self):
+        # set up interface to adapter properties using high-level api
+        bus = await MessageBus(bus_type=BusType.SYSTEM).connect()
+        introspect = await bus.introspect(BLUEZ, BLUEZ_PATH)
+        obj = bus.get_proxy_object(BLUEZ, BLUEZ_PATH, introspect)
+        iface = obj.get_interface(BLUEZ_ADAPTER)
+        props = obj.get_interface(BLUEZ_PROPERTIES)
+
+        powered = await iface.get_powered()
+        # subscribe receiver to property changed
+        props.on_properties_changed(self._signal_received_adapter)
+        return powered
+
+    async def _init_device(self):
+        # set up interface to device properties using high-level api
+        bus = await MessageBus(bus_type=BusType.SYSTEM).connect()
+        introspect = await bus.introspect(BLUEZ, BLUEZ_PATH + self.hci)
+        obj = bus.get_proxy_object(BLUEZ, BLUEZ_PATH + self.hci, introspect)
+        iface = obj.get_interface(BLUEZ_DEVICE)
+        props = obj.get_interface(BLUEZ_PROPERTIES)
+
+        connected = await iface.get_connected()
+        name = await iface.get_name()
+        # subscribe receiver to property changed
+        props.on_properties_changed(self._signal_received_device)
+        return connected, name
+
+    def _signal_received_adapter(self, interface_name, changed_properties, invalidated_properties):
+        powered = changed_properties.get('Powered', None)
+        if powered is not None:
+            self.powered = powered.value
+            self.update_text()
+
+    def _signal_received_device(self, interface_name, changed_properties, invalidated_properties):
+        connected = changed_properties.get('Connected', None)
+        if connected is not None:
+            self.connected = connected.value
+            self.update_text()
+
+        device = changed_properties.get('Name', None)
+        if device is not None:
+            self.device = device.value
+            self.update_text()
+
+    def update_text(self):
+        if not self.powered:
+            self.text = 'off'
```

Comment:
Rather than hard-coding the `on` and `off` text, you could set these via the config (but have the defaults be as you have them now).